Shuttle manipulator

coiax opened the pull request:
#16761 is in a stable but crude state.

This PR is to add a more streamlined interface for admins to select shuttle templates instead of just crudely hitting verbs.
- Removes shuttle verbs.
- Gives features to shuttle manipulator at Centcom.
- Look at a list of shuttle templates that are appropriate for the one you want to replace.
- Also monitor all the shuttles I guess.
- It allows hotswapping of shuttles while they're recalling or being summoned.
- Also some GC changes to allow us to hard delete items like greentext or mobile docking ports.

:cl:
rscdel: Emagging the emergency shuttle computer with less than ten seconds until launch no longer gives additional time.
rscadd: The emergency shuttle computer now reads the ID in your ID slot, rather than your hand.
/:cl:
- Added a new special statue
- All shuttle templates are now "timid" and will not register themselves with the shuttles subsystem on load
- Barshuttle now has a Barmaiden and anti greytide tables
- All items that return QDEL_HINT_LETMELIVE must also take the force parameter, and call super and return it if force = TRUE.
- Shuttles now give visible messages when road killing mobs
### Interface

![image](https://cloud.githubusercontent.com/assets/609465/15450667/ab4d31da-1f99-11e6-80b6-7192601d253b.png)
![image](https://cloud.githubusercontent.com/assets/609465/15450671/c4670be6-1f99-11e6-8cff-15b94b5ddf58.png)
![image](https://cloud.githubusercontent.com/assets/609465/15450676/daf6020e-1f99-11e6-93bd-7038e511c542.png)

coiax committed:
Starting out on our tgui journey

Dorsisdwarf commented:
Mother of chrome, the comits

TrustyGun commented:
![untitled](https://cloud.githubusercontent.com/assets/7079721/15087844/d8c5ef5c-13ba-11e6-83bc-4682991c6aa0.png)

coiax commented:
73 of these commits are in the shuttles PR, which I expect to be merged within the week. And it's been actively worked on, I don't know why so many commits are bad.

TrustyGun commented:
its not bad, i just found it funny how i had to zoom out 3 times for all of them to show

Durkel commented:
I guess you could say coiax is commited to the pr.

PKPenguin321 commented:
coiax do you not know that you can branch off of master?

coiax commented:
I'm making a PR that requires another PR to be merged. I can't branch off on master.

coiax committed:
God tgui, why do you need to be updated

You're dynamically generated anyway

Bawhoppen commented:
coiax why you weren't supposed to make a pr yet did you even listen to me 
reeeee

coiax commented:
@Bawhoppen, this isn't your player made shuttles, this is just an extension of my shuttle stuff, which I'm doing in a new branch so the existing stuff gets merged

Bawhoppen commented:
I assumed it was part of it is what I meant

coiax committed:
Commit of stuff

coiax committed:
Further progress is being made

coiax committed:
Everyone loves buildscripts

coiax committed:
Further modifications to my incredible running script

coiax committed:
Starting to modify the minimap code to be better

coiax committed:
It's going well thusfar, I guess

coiax committed:
What have I done

coiax committed:
RIP minimap

coiax committed:
FUN FUN FUN FUN FUN

coiax committed:
Adds shuttle_manipulator.dmi

MUH HOLOGRAMS

neersighted commented:
you are a brave man. ping me when you're ready for review

coiax committed:
Merge branch 'master' into shuttle-manipulator

Conflicts:
	tgui/assets/tgui.css
	tgui/assets/tgui.js

coiax committed:
Is it done? IS IT OVER

coiax commented:
@neersighted ping ping review

neersighted commented on code/modules/shuttle/manipulator.dm:
```diff
@@ -9,5 +9,214 @@
+		L["description"] = S.description
+		L["admin_notes"] = S.admin_notes
+
+		if(selected && (selected.shuttle_id == S.shuttle_id))
```
I'd prefer that this be a flag of some sort... Maybe have selected be the name? Sending data down the wire twice when it can be derived is wasteful.

(a computed variable could easily be used)

neersighted commented on code/modules/shuttle/manipulator.dm:
```diff
@@ -9,5 +9,214 @@
+
+		templates[S.port_id]["templates"] += list(L)
+
+	data["templates_tabs"] = sortList(data["templates_tabs"])
```
It's clearer to use a temporary var first and then assign to data, but it's up to your preference

neersighted commented on code/modules/shuttle/manipulator.dm:
```diff
@@ -9,5 +9,214 @@
+	// - removed old shuttle
+	// - moved new shuttle to old location
+
+	//world.log << json_encode(data)
```
delete :p

neersighted commented on code/modules/shuttle/manipulator.dm:
```diff
@@ -9,5 +9,214 @@
+
+			var/datum/map_template/shuttle/S = shuttle_templates[shuttle_id]
+			if(S)
+				src.selected = S
```
Don't use src

neersighted commented on code/modules/shuttle/manipulator.dm:
```diff
@@ -9,5 +9,214 @@
+				for(var/i in SSshuttle.mobile)
+					var/obj/docking_port/mobile/M = i
+					if(M.id == params["id"])
+						user.loc = get_turf(M)
```
I think we use forceMove instead of setting loc now

neersighted commented on code/modules/shuttle/manipulator.dm:
```diff
@@ -9,5 +9,214 @@
+				load_template(S)
+				if(preview_shuttle)
+					preview_shuttle_id = shuttle_id
+					user.loc = get_turf(preview_shuttle)
```
Ditto

neersighted commented on code/modules/shuttle/manipulator.dm:
```diff
@@ -9,5 +9,214 @@
+			var/shuttle_id = params["id"]
+			var/datum/map_template/shuttle/S = shuttle_templates[shuttle_id]
+			if(S)
+				if(!preview_shuttle || (S.shuttle_id != preview_shuttle_id))
```
Tidy up this control flow?

neersighted commented on code/modules/shuttle/manipulator.dm:
```diff
@@ -9,5 +9,214 @@
+				var/new_id = replacetext(preview_shuttle.id, "(preview)", "")
+				preview_shuttle.id = new_id
+
+				preview_shuttle.dock(D)
```
More comments would be good

neersighted commented on code/modules/shuttle/manipulator.dm:
```diff
@@ -9,5 +9,214 @@
+	update_icon()
+
+
+/obj/machinery/shuttle_manipulator/proc/load_template(
```
Remove linebreak

coiax replied:
muh 79 character lines

neersighted commented on code/modules/shuttle/manipulator.dm:
```diff
@@ -9,5 +9,214 @@
+				existing_shuttle = null
+				selected = null
+
+	update_icon()
```
Pretty sure there is no icon? maybe only update if `. = TRUE` (ie UI needs updating)?

neersighted commented on code/modules/tgui/states.dm:
```diff
@@ -16,14 +16,20 @@
  **/
 /datum/proc/ui_status(mob/user, datum/ui_state/state)
 	var/src_object = ui_host()
+	. = UI_CLOSE
```
I'm guessing the purpose of this is to allow `admin_state` when dead? Fair enough, that was an oversight when I added the ghost interaction.

coiax replied:
pretty much. It was a weird bug until I figured it out.

neersighted commented on run.py:
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python
```
What does this have to do with the PR?

neersighted commented on tgui/assets/tgui.css:
```diff
@@ -1 +1,924 @@
-@charset "utf-8";body,html{box-sizing:border-box;height:100%;margin:0}html{overflow:hidden;cursor:default}body{overflow:auto;font-family:Verdana,Geneva,sans-serif;font-size:12px;color:#fff;background-color:#2a2a2a;background-image:linear-gradient(180deg,#2a2a2a 0,#202020);filter:progid:DXImageTransform.Microsoft.gradient(startColorstr='#ff2a2a2a',endColorstr='#ff202020',GradientType=0)}*,:after,: ...
\ No newline at end of file
+@charset "utf-8";
+html,
```
Assets need to be compiled with `--min`.

neersighted commented on tgui/src/interfaces/shuttle_manipulator/templates.ract:
```diff
@@ -0,0 +1,23 @@
+						<ui-section label='Admin Notes'>{{admin_notes}}</ui-section>
+					{{/if}}
+
+<ui-button action='select_template'
```
Clean the indentation please (spaces, two or four. we like two mostly)

neersighted commented on tgui/src/interfaces/shuttle_manipulator/status.ract:
```diff
@@ -0,0 +1,11 @@
+			{{#if timer}}
+				({{timeleft}})
+			{{/if}}
+<ui-button action='jump_to'
```
Clean the indentation

neersighted commented on tgui/src/interfaces/shuttle_manipulator/modification.ract:
```diff
@@ -0,0 +1,35 @@
+{{/if}}
+{{#if !data.existing_shuttle}}
+<ui-display title='Existing Shuttle: None'></ui-display>
+{{/if}}
```
Indentation helps readability

neersighted commented on tgui/src/interfaces/shuttle_manipulator.ract:
```diff
@@ -0,0 +1,20 @@
+
+<ui-tabs tabs='{{data.tabs}}'>
+	<tab name='Status'>
+		<status></status>
```
`<status/>`

neersighted commented on tgui/src/interfaces/shuttle_manipulator.ract:
```diff
@@ -0,0 +1,20 @@
+		<status></status>
+	</tab>
+	<tab name='Templates'>
+		<templates></templates>
```
ditto

neersighted commented on tgui/src/interfaces/shuttle_manipulator.ract:
```diff
@@ -0,0 +1,20 @@
+	</tab>
+	<tab name='Modification'>
+		{{#if data.selected}}
+		<modification></modification>
```
ditto

neersighted commented on tgui/src/interfaces/shuttle_manipulator.ract:
```diff
@@ -0,0 +1,20 @@
+		<modification></modification>
+		{{/if}}
+		{{#if !data.selected}}
+		No shuttle selected.
```
In a `<span class='bad'>` please?

neersighted commented on tgui/build.sh:
```diff
@@ -0,0 +1,11 @@
+#RUN THIS IN THE tgui/ folder
+#!/bin/bash
```
Should be the first line on older kernels

neersighted commented on tgui/build.sh:
```diff
@@ -0,0 +1,11 @@
+export NODE_VERSION=4
+
+if [ ! -d "/tmp/nvm" ]; then
+	rm -rf /tmp/nvm && git clone https://github.com/creationix/nvm.git /tmp/nvm && (cd ~/.nvm && git checkout `git describe --abbrev=0 --tags`) && source /tmp/nvm/nvm.sh && nvm install $NODE_VERSION
```
Not thrilled with this script but oh well, supporting linux development is a good idea

neersighted commented on tgui/build.sh:
```diff
@@ -0,0 +1,11 @@
+fi
+
+source /tmp/nvm/nvm.sh
+gulp
```
Just don't forget about gulp watch and `--min` before commit

neersighted commented on code/modules/shuttle/shuttle.dm:
```diff
@@ -18,18 +18,16 @@
+/obj/docking_port/Destroy(force)
 	// unless you assert that you know what you're doing. Horrible things
 	// may result.
-	if(!i_know_what_im_doing)
```
classy

neersighted commented on code/modules/shuttle/shuttle.dm:
```diff
@@ -346,25 +346,13 @@
-
-	src.i_know_what_im_doing = TRUE
-	qdel(src)
+	qdel(src, force=TRUE)
```
thank you for using named parameters

neersighted commented on code/modules/shuttle/manipulator.dm:
```diff
@@ -9,5 +9,214 @@
+/obj/machinery/shuttle_manipulator/proc/load_template(
+	datum/map_template/shuttle/S)
+	// load shuttle template, centred at shuttle import landmark,
+	var/turf/landmark_turf = get_turf(locate("landmark*Shuttle Import"))
```
magic strings aaaaaaa

Iamgoofball replied:
thats the landmark system we cant really do much

neersighted commented:
heyoooo

coiax committed:
Peer review

coiax committed:
Some bug fixes

coiax committed:
Makes that damn greentext shut up

coiax committed:
Shuttle registration

coiax committed:
Made the Emergency Escape Bar more robust

No climbing on the bar.

coiax committed:
Do not stare into the operation end of the device

coiax committed:
Compile shame

coiax committed:
THE MOST DUMB

coiax committed:
Passive barmaiden

coiax committed:
Emergency shuttle memes

coiax committed:
MORE SAFETY CODE

coiax committed:
Fancy shuttle manipulator icons

coiax committed:
Smoothing it out

coiax committed:
We are going to have a lot of fun with this one

coiax committed:
Merge branch 'master' into shuttle-manipulator

Conflicts:
	code/__DEFINES/misc.dm
	code/modules/awaymissions/mission_code/challenge.dm

coiax committed:
Independent blobbernauts

coiax committed:
WABBAJACK WABBAJACK

Shadowlight213 commented on code/game/gamemodes/blob/blobs/blob_mobs.dm:
```diff
@@ -271,3 +275,7 @@
 		factory.maxhealth = initial(factory.maxhealth)
 	flick("blobbernaut_death", src)
+
+/mob/living/simple_animal/hostile/blob/blobbernaut/independent
```
How is this related to shuttles? Or am i missing something here?

coiax replied:
It has nothing to do with shuttles, but I discovered that a bolt of changed blobbernaut would start dying while making and testing the wabbajack shuttle.

Shadowlight213 replied:
Atomize your prs meme

Iamgoofball replied:
nope, this would be a bugfix for the wabbajack

ChangelingRain replied:
Atomize the prs! Shove this into the polymorph pr you already have open.

coiax committed:
Merge branch 'master' into shuttle-manipulator

Conflicts:
	tgui/assets/tgui.js

coiax committed:
Message for attempting to authenticate old style

coiax committed:
Merge branch 'master' into shuttle-manipulator

Conflicts:
	tgui/assets/tgui.js

coiax committed:
Angry alert noise is back

coiax committed:
Revert "Independent blobbernauts"

This reverts commit 34d6af7c9c88cfc2864990cb37b586bb90163dd3.

coiax committed:
Merge branch 'master' into shuttle-manipulator

Conflicts:
	_maps/shuttles/emergency_meta.dmm

coiax committed:
No parrot sleep plz

RemieRichards commented on code/modules/mob/mob_helpers.dm:
```diff
@@ -272,6 +272,10 @@ It's fairly easy to fix if dealing with single letters but not so much with comp
 			return M
 	return 0
 
+/proc/random_name()
```
we already have one of these.

RemieRichards commented on code/modules/shuttle/manipulator.dm:
```diff
@@ -6,8 +6,267 @@
+			master_ui, state)
+		ui.open()
+
+/proc/shuttlemode2str(mode)
```
this shouldn't exist, make the defines themselves strings.

RemieRichards commented on code/modules/shuttle/emergency.dm:
```diff
@@ -6,61 +11,164 @@
+	data["engines_started"] = ENGINES_STARTED
+	data["authorizations_remaining"] = max((auth_need - authorized.len), 0)
+	var/A = list()
+	data["authorizations"] = A
```
this weirds me out so much.
I know lists are objects.
I know they're pass by reference.
but you're setting data before it's been added and it really just irks me.

RemieRichards commented on code/modules/shuttle/emergency.dm:
```diff
@@ -6,61 +11,164 @@
+	data["timer_str"] = SSshuttle.emergency.getTimerStr()
+	data["engines_started"] = ENGINES_STARTED
+	data["authorizations_remaining"] = max((auth_need - authorized.len), 0)
+	var/A = list()
```
this is a list, so typecast it.

RemieRichards commented on code/game/objects/structures/tables_racks.dm:
```diff
@@ -325,18 +325,46 @@
+	else
+		. = ..()
+
+/obj/structure/table/wood/bar/proc/is_barstaff(mob/user)
```
as said in IRC, check for bar access on the ID instead of this istype stuff.

coiax committed:
Moves the special shuttle stuff to special.dm

coiax committed:
No Bartender like a Centcom Bartender

coiax committed:
Merge branch 'master' into shuttle-manipulator

coiax committed:
Non-controversial map changes

- Backup shuttle moved closer to Centcom main structure, docking tube
created
- Moved shuttle import landmark to above Centcom main building
- Added shuttle displays to Conference Room
- Squashed the Chapel a bit in Metastation
- Made the docking port on Z2 massive
- Made the docking port on Metastation a lot larger

coiax committed:
Hacks and slashes at Box

A bunch of things are extended and squashed so Box shuttle dock can
support the MetaStation emergency shuttle.

coiax committed:
Some Metastationshit

coiax committed:
Never ending changes

coiax committed:
Wabbajack to TGM

coiax committed:
Modified the bar, I think that's all of them

coiax committed:
Merge branch 'master' into shuttle-manipulator

coiax committed:
Stops Barmaiden wandering around

coiax committed:
More code review

coiax committed:
Whitspace, the bane of us all

coiax committed:
DIE WHITESPACE DIE

Bluespace commented:
Fun fact, this PR made rounds end instantly!

Dorsisdwarf commented:
Nice
